okhttp-clients/src/main/java/com/palantir/remoting3/okhttp/RemotingOkHttpCall.java

```diff
@@ -0,0 +1,238 @@
+                                callback.onFailure(call, new IOException("Exceeded the maximum number of allowed "
+                                        + "redirects for initial URL: " + call.request().url()));
+                            } else {
+                                Optional<HttpUrl> redirectTo = urls.redirectTo(request().url(),
+                                        exception.getRedirectTo().toString());
+                                if (!redirectTo.isPresent()) {
+                                    callback.onFailure(call, new IOException("Failed to determine valid redirect URL "
+                                            + "for '" + exception.getRedirectTo() + "' and base URLs "
+                                            + urls.getBaseUrls()));
+                                } else {
+                                    Request redirectedRequest = request().newBuilder()
+                                            .url(redirectTo.get())
+                                            .build();
+                                    client.newCallWithMutableState(
+                                            redirectedRequest, backoffStrategy, maxNumRelocations - 1)
+                                            .enqueue(callback);
+                                }
+                            }
+                            return null;
+                        }
+
+                        @Override
+                        public Void visit(QosException.Unavailable exception) {
+                            Optional<Duration> backoff = backoffStrategy.nextBackoff();
+                            if (!backoff.isPresent()) {
+                                log.debug("Max number of retries exceeded, failing call");
+                                callback.onFailure(call, new IOException("Failed to complete the request due to a "
+                                        + "server-side QoS condition: 503", exception));
+                            } else {
+                                log.debug("Rescheduling call after backoff",
+                                        SafeArg.of("backoffMillis", backoff.get().toMillis()));
+                                scheduleExecution(callback, backoff.get());
+                            }
+                            return null;
+                        }
+                    });
+                    return;
+                }
+
+                // Handle responses that correspond to RemoteExceptions / SerializableErrors
+                Optional<RemoteException> httpError = remoteExceptionHandler.handle(response);
+                if (httpError.isPresent()) {
+                    callback.onFailure(call, new IoRemoteException(httpError.get()));
+                    return;
+                }
+
+                // Catch-all: handle all other responses
+                Optional<IOException> ioException = catchAllExceptionHandler.handle(response);
```

Comment:
Seems a bit surprising that the 'catchAllExceptionHandler' doesn't actually catch everything and lets some stuff fall through to that last case.  Can we rename/change behaviour to make this more obvious?

Reply:
Do you have a proposal for how to do this? The handlers basically implement a switch statements over HTTP codes and at the bottom is the `default` case. Not sure how to check at compile-time that these handlers are exhaustive.

Reply:
We could rename this handler to say `catchEverythingExceptionHttp101` so that the name is consistent with it's signature, or if we want to keep it named `catchAll`, maybe we should just make this specific one not implement the `ResponseHandler` interface and just always return an IOException?

As an aside, what is the desired behaviour for a http 101 code?  At the moment it'll zoom all the way down the bottom here and trigger `new IOException("Failed to handle request, this is an http-remoting bug.")`.  Do remoting-servers return 101?

Reply:
a 101 should really never make it to this stage.

Reply:
okhttp internals should catch it and trigger a protocol upgrade

Reply:
@jared2501 , is that correct?

Reply:
will now also relay 1xx responses as successful.

Reply:
renamed exception handler